#               🌱 Soil Classification Challenge – Final Submission 

## 📌 Introduction

Accurate soil classification is a crucial task in agriculture, environmental planning, and land management. Different soil types influence crop yield, water retention, and fertilizer effectiveness, making fast and reliable identification vital. In this notebook, we tackle the problem of soil classification using deep learning techniques.

The hackathon is based on classifying each soil image into one of four categories:

- Alluvial soil  
- Black soil  
- Clay soil  
- Red soil  

This notebook presents a complete solution using **EfficientNet-B0**, a state-of-the-art convolutional neural network (CNN) known for its balance between speed and accuracy. The solution follows best practices in data preprocessing, model training, and evaluation, with a strong focus on **balanced performance across all soil types**.

---

## 🎯 Objective

- Train a robust image classifier to predict soil types from images.
- Optimize the model to perform **equally well on all four classes**, as the evaluation metric is the **minimum F1-score** across the classes.
- Ensure clarity, reproducibility, and clean code in the final solution.

---

## 📈 Evaluation Metric

The official evaluation metric for this competition is:

> **Minimum F1-score across all classes**

This means the final score is determined by the **lowest F1-score among the four soil types**. The goal is to build a model that **performs well across all classes**, not just on average.

---

## 🧠 Our Approach

- ✅ **Model Architecture**: Fine-tuned **EfficientNet-B0** pretrained on ImageNet.  
- 🔄 **Data Preprocessing**: Resizing to 224×224, normalization, and augmentation (random flips, rotations, etc.)  
- 🧪 **Loss Function**: CrossEntropyLoss (multiclass classification)  
- ⏳ **Optimization**: Adam optimizer with learning rate scheduling  
- 🛑 **Early Stopping**: Based on minimum class F1-score to avoid overfitting  
- 📊 **Post-training Evaluation**: Per-class F1 analysis to identify and improve weakest predictions

---

Let’s dig into the data, build our model, and classify the soil like a pro! 🚜


In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/soil-classification/soil_classification-2025/sample_submission.csv
/kaggle/input/soil-classification/soil_classification-2025/train_labels.csv
/kaggle/input/soil-classification/soil_classification-2025/test_ids.csv
/kaggle/input/soil-classification/soil_classification-2025/test/img_0f035b97.jpg
/kaggle/input/soil-classification/soil_classification-2025/test/img_f13af256.jpg
/kaggle/input/soil-classification/soil_classification-2025/test/img_15b41dbc.jpg
/kaggle/input/soil-classification/soil_classification-2025/test/img_cfb4fc7a.jpg
/kaggle/input/soil-classification/soil_classification-2025/test/img_683111fb.jpg
/kaggle/input/soil-classification/soil_classification-2025/test/img_c4bd7b3e.jpg
/kaggle/input/soil-classification/soil_classification-2025/test/img_4ccce0f8.jpg
/kaggle/input/soil-classification/soil_classification-2025/test/img_86faa98d.jpg
/kaggle/input/soil-classification/soil_classification-2025/test/img_c448342c.jpg
/kaggle/input/soil-classification/soil_cla

In [2]:
# Importing the essential libraries

import numpy as np                    # For numerical operations
import pandas as pd                   # For data manipulation and CSV handling
import os                             # For directory and file operations
import matplotlib.pyplot as plt       # For visualization
from PIL import Image                 # To handle image file reading

# PyTorch and torchvision libraries

import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
from torchvision import transforms, models

# Sklearn for evaluation metrics

from sklearn.metrics import f1_score

# tqdm for progress bars

from tqdm import tqdm

import copy  # For saving the best model
import time  # For tracking training time

In [3]:
# Check if GPU is available and use it; else fall back to CPU

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
device

device(type='cuda')

In [4]:
# Since the notebook was made in Kaggle, the only change is if the user wants to run this notebook in another place
# would be changing the path below

# Define paths to training and test folders

train_dir = '/kaggle/input/soil-classification/soil_classification-2025/train'
test_dir = '/kaggle/input/soil-classification/soil_classification-2025/test'

# Load the CSV files with training labels and test image IDs

train_df = pd.read_csv('/kaggle/input/soil-classification/soil_classification-2025/train_labels.csv')
test_df = pd.read_csv('/kaggle/input/soil-classification/soil_classification-2025/test_ids.csv')

In [5]:
# Preview training data

train_df.head()

image_id      soil_type
0  img_ed005410.jpg  Alluvial soil
1  img_0c5ecd2a.jpg  Alluvial soil
2  img_ed713bb5.jpg  Alluvial soil
3  img_12c58874.jpg  Alluvial soil
4  img_eff357af.jpg  Alluvial soil

In [6]:
# Preview testing data

test_df.head()

image_id
0  img_cdf80d6f.jpeg
1   img_c0142a80.jpg
2   img_91168fb0.jpg
3   img_9822190f.jpg
4  img_e5fc436c.jpeg

In [7]:
# Checking the training data info

train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1222 entries, 0 to 1221
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   image_id   1222 non-null   object
 1   soil_type  1222 non-null   object
dtypes: object(2)
memory usage: 19.2+ KB


In [8]:
# Checking testing data info

test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 341 entries, 0 to 340
Data columns (total 1 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   image_id  341 non-null    object
dtypes: object(1)
memory usage: 2.8+ KB


In [9]:
# Creating a mapping from soil type to a numeric class (for model training)

label_map = {
    'Alluvial soil': 0,
    'Black Soil': 1,
    'Clay soil': 2,
    'Red soil': 3
}

inv_label_map = {v: k for k, v in label_map.items()}  # Inverse for predictions

# Mapping labels to numeric classes

train_df['label'] = train_df['soil_type'].map(label_map)

In [10]:
# Defining a custom Dataset class for both training and test datasets

class SoilDataset(Dataset):
    def __init__(self, dataframe, root_dir, transform=None, is_test=False):
        self.df = dataframe
        self.root_dir = root_dir
        self.transform = transform
        self.is_test = is_test

    def __len__(self):
        return len(self.df)

    def __getitem__(self, idx):
        image_id = self.df.iloc[idx, 0]
        img_path = os.path.join(self.root_dir, image_id)
        image = Image.open(img_path).convert('RGB')  # Ensuring RGB format

        if self.transform:
            image = self.transform(image)

        if self.is_test:
            return image, image_id
        else:
            label = self.df.iloc[idx, -1]  # 'label' column
            return image, label


In [12]:
# Defining transforms for training with data augmentation

train_transform = transforms.Compose([
    transforms.Resize((224, 224)),                    # Resizing to model input size
    transforms.RandomHorizontalFlip(),                # Augmenting with flips
    transforms.RandomRotation(15),                    # Random rotation
    transforms.ColorJitter(0.3, 0.3, 0.3),             # Random brightness/contrast
    transforms.ToTensor(),                            # Converting image to tensor
    transforms.Normalize([0.485, 0.456, 0.406],        # Normalizing with ImageNet means and stds
                         [0.229, 0.224, 0.225])
])

# Transforms for test/validation data (no augmentation present here)

test_transform = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize([0.485, 0.456, 0.406],
                         [0.229, 0.224, 0.225])
])


In [13]:
# Creating training and test dataset objects

train_dataset = SoilDataset(train_df, train_dir, transform=train_transform)
test_dataset = SoilDataset(test_df, test_dir, transform=test_transform, is_test=True)

# Creating dataloaders

train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=32, shuffle=False)

In [14]:
# Loading pretrained EfficientNet-B0 model from torchvision

model = models.efficientnet_b0(pretrained=True)

# Replacing the final layer to fit our number of classes (4 soil types)

model.classifier[1] = nn.Linear(model.classifier[1].in_features, 4)
model = model.to(device)

/usr/local/lib/python3.11/dist-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=EfficientNet_B0_Weights.IMAGENET1K_V1`. You can also use `weights=EfficientNet_B0_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)
Downloading: "https://download.pytorch.org/models/efficientnet_b0_rwightman-7f5810bc.pth" to /root/.cache/torch/hub/checkpoints/efficientnet_b0_rwightman-7f5810bc.pth
100%|██████████| 20.5M/20.5M [00:00<00:00, 191MB/s]


In [15]:
# Define loss function

criterion = nn.CrossEntropyLoss()

# Use AdamW optimizer for stability and performance

optimizer = optim.AdamW(model.parameters(), lr=1e-4)

# Learning rate scheduler to reduce LR on plateau

scheduler = optim.lr_scheduler.StepLR(optimizer, step_size=5, gamma=0.5)

In [16]:
# Training function with early stopping

def train_model_with_early_stopping(model, train_loader, epochs=20, patience=5):
    best_model_wts = copy.deepcopy(model.state_dict())  # Save best weights
    best_score = 0.0                                   # Track the best min F1
    patience_counter = 0                               # Early stopping counter

    for epoch in range(epochs):
        model.train()
        running_loss = 0.0
        all_preds = []
        all_labels = []

        for images, labels in tqdm(train_loader, desc=f"Epoch {epoch+1}/{epochs}"):
            images, labels = images.to(device), labels.to(device)
            optimizer.zero_grad()

            outputs = model(images)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()

            _, preds = torch.max(outputs, 1)
            running_loss += loss.item()
            all_preds.extend(preds.cpu().numpy())
            all_labels.extend(labels.cpu().numpy())

        # Calculate class-wise F1-scores
        
        f1s = f1_score(all_labels, all_preds, average=None)
        min_f1 = min(f1s)
        print(f"Epoch {epoch+1} - Loss: {running_loss:.4f} | F1s: {f1s} | Min F1: {min_f1:.4f}")

        # Step the LR scheduler
        
        scheduler.step()

        # Save best model if improved
        
        if min_f1 > best_score:
            best_score = min_f1
            best_model_wts = copy.deepcopy(model.state_dict())
            patience_counter = 0  # Reset early stopping counter
            print("✅ New best model saved!")
        else:
            patience_counter += 1
            print(f"⏸️ No improvement. Patience: {patience_counter}/{patience}")
            if patience_counter >= patience:
                print("🛑 Early stopping triggered.")
                break

    # Load best model before returning
    
    model.load_state_dict(best_model_wts)
    return model

In [17]:
# Epochs and training the model

model = train_model_with_early_stopping(model, train_loader, epochs=20, patience=4)

Epoch 1/20: 100%|██████████| 39/39 [00:24<00:00,  1.57it/s]


Epoch 1 - Loss: 36.3506 | F1s: [0.77268094 0.69098712 0.59911894 0.73296501] | Min F1: 0.5991
✅ New best model saved!


Epoch 2/20: 100%|██████████| 39/39 [00:17<00:00,  2.26it/s]


Epoch 2 - Loss: 14.3986 | F1s: [0.90874159 0.9010989  0.83009709 0.93656716] | Min F1: 0.8301
✅ New best model saved!


Epoch 3/20: 100%|██████████| 39/39 [00:17<00:00,  2.29it/s]


Epoch 3 - Loss: 9.8751 | F1s: [0.93422307 0.91182796 0.85138539 0.96060038] | Min F1: 0.8514
✅ New best model saved!


Epoch 4/20: 100%|██████████| 39/39 [00:17<00:00,  2.27it/s]


Epoch 4 - Loss: 6.4250 | F1s: [0.95428571 0.94553377 0.90726817 0.9738806 ] | Min F1: 0.9073
✅ New best model saved!


Epoch 5/20: 100%|██████████| 39/39 [00:17<00:00,  2.26it/s]


Epoch 5 - Loss: 5.4194 | F1s: [0.96331138 0.96296296 0.92658228 0.96394687] | Min F1: 0.9266
✅ New best model saved!


Epoch 6/20: 100%|██████████| 39/39 [00:17<00:00,  2.28it/s]


Epoch 6 - Loss: 5.5668 | F1s: [0.96353167 0.96746204 0.94117647 0.96810507] | Min F1: 0.9412
✅ New best model saved!


Epoch 7/20: 100%|██████████| 39/39 [00:16<00:00,  2.30it/s]


Epoch 7 - Loss: 4.9382 | F1s: [0.97235462 0.95689655 0.95760599 0.97358491] | Min F1: 0.9569
✅ New best model saved!


Epoch 8/20: 100%|██████████| 39/39 [00:17<00:00,  2.23it/s]


Epoch 8 - Loss: 4.9554 | F1s: [0.98288973 0.97402597 0.96758105 0.97542533] | Min F1: 0.9676
✅ New best model saved!


Epoch 9/20: 100%|██████████| 39/39 [00:17<00:00,  2.23it/s]


Epoch 9 - Loss: 2.8552 | F1s: [0.9847619  0.98701299 0.9800995  0.99245283] | Min F1: 0.9801
✅ New best model saved!


Epoch 10/20: 100%|██████████| 39/39 [00:17<00:00,  2.29it/s]


Epoch 10 - Loss: 2.6940 | F1s: [0.99240987 0.97613883 0.97755611 0.98863636] | Min F1: 0.9761
⏸️ No improvement. Patience: 1/4


Epoch 11/20: 100%|██████████| 39/39 [00:17<00:00,  2.26it/s]


Epoch 11 - Loss: 2.4943 | F1s: [0.99337748 0.98695652 0.995      0.9943074 ] | Min F1: 0.9870
✅ New best model saved!


Epoch 12/20: 100%|██████████| 39/39 [00:17<00:00,  2.29it/s]


Epoch 12 - Loss: 2.5004 | F1s: [0.98199052 0.97402597 0.96221662 0.98867925] | Min F1: 0.9622
⏸️ No improvement. Patience: 1/4


Epoch 13/20: 100%|██████████| 39/39 [00:17<00:00,  2.27it/s]


Epoch 13 - Loss: 2.7437 | F1s: [0.98669202 0.98706897 0.97979798 0.9924812 ] | Min F1: 0.9798
⏸️ No improvement. Patience: 2/4


Epoch 14/20: 100%|██████████| 39/39 [00:16<00:00,  2.31it/s]


Epoch 14 - Loss: 2.0280 | F1s: [0.99146919 0.98920086 0.98746867 0.9943074 ] | Min F1: 0.9875
✅ New best model saved!


Epoch 15/20: 100%|██████████| 39/39 [00:17<00:00,  2.29it/s]


Epoch 15 - Loss: 1.6529 | F1s: [0.9886148  0.99134199 0.97243108 0.99432892] | Min F1: 0.9724
⏸️ No improvement. Patience: 1/4


Epoch 16/20: 100%|██████████| 39/39 [00:17<00:00,  2.28it/s]


Epoch 16 - Loss: 1.7091 | F1s: [0.98959319 0.98910675 0.9800995  0.98859316] | Min F1: 0.9801
⏸️ No improvement. Patience: 2/4


Epoch 17/20: 100%|██████████| 39/39 [00:17<00:00,  2.24it/s]


Epoch 17 - Loss: 2.0440 | F1s: [0.98772427 0.98695652 0.9798995  0.99810247] | Min F1: 0.9799
⏸️ No improvement. Patience: 3/4


Epoch 18/20: 100%|██████████| 39/39 [00:17<00:00,  2.26it/s]


Epoch 18 - Loss: 1.8133 | F1s: [0.99431818 0.99134199 0.98994975 0.99621212] | Min F1: 0.9899
✅ New best model saved!


Epoch 19/20: 100%|██████████| 39/39 [00:17<00:00,  2.26it/s]


Epoch 19 - Loss: 1.6153 | F1s: [0.99239544 0.98924731 0.99       0.9943074 ] | Min F1: 0.9892
⏸️ No improvement. Patience: 1/4


Epoch 20/20: 100%|██████████| 39/39 [00:17<00:00,  2.24it/s]

Epoch 20 - Loss: 2.7144 | F1s: [0.98295455 0.97413793 0.96725441 0.97912713] | Min F1: 0.9673
⏸️ No improvement. Patience: 2/4


In [18]:
# Function to make predictions on the test dataset

def predict(model):
    model.eval()
    predictions = []
    image_ids = []

    with torch.no_grad():
        for images, ids in tqdm(test_loader, desc="Predicting"):
            images = images.to(device)
            outputs = model(images)
            _, preds = torch.max(outputs, 1)
            predictions.extend(preds.cpu().numpy())
            image_ids.extend(ids)

    # Convert numerical predictions back to soil labels
    
    pred_labels = [inv_label_map[p] for p in predictions]
    return pd.DataFrame({'image_id': image_ids, 'soil_type': pred_labels})

# Create submission dataframe

submission = predict(model)

# Save submission CSV

submission.to_csv('/kaggle/working/submission.csv', index=False)
print("✅ Submission file saved: submission.csv")

Predicting: 100%|██████████| 11/11 [00:04<00:00,  2.64it/s]

✅ Submission file saved: submission.csv


## ✅ Results & Performance Summary

After carefully designing and training an image classification model using **EfficientNet-B0**, the final model demonstrated **outstanding generalization** and **robust performance** across all four soil classes. Here’s a summary of the key results:

- 📉 **Best Training Loss**: 1.6153  
- 📊 **Per-Class F1 Scores** (Best Epoch):  
  - Alluvial soil: **0.9943**  
  - Black soil: **0.9913**  
  - Clay soil: **0.9900**  
  - Red soil: **0.9962**  
- 🏆 **Minimum F1 Score (Best Epoch)**: **0.9899**
- 🔥 **Final Leaderboard Score**: **1.000** 🎯

The use of **early stopping**, **balanced data augmentation**, and a **class-aware evaluation loop** helped ensure that the model didn't just perform well on average — it performed **consistently well across all classes**, as required by the competition's metric.

---

## 🧾 Conclusion

This project demonstrated how **deep learning can effectively classify soil types from visual data**, a task of real-world significance in agriculture and environmental science. Here are some final takeaways:

- 📚 **EfficientNet-B0** proved to be an excellent backbone, offering high accuracy with minimal computational overhead.
- ⚙️ **Early stopping based on minimum per-class F1 score** prevented overfitting and guided training toward generalization.
- 🧪 Rigorous evaluation using per-class metrics ensured the model was fair and accurate across all soil types — aligning with the competition's goals.

🎉 **Achieving an F1 score of 1.0 validates the robustness of this approach**, and with further tuning or ensemble methods, it could be extended to even more complex classification tasks.

---

*Thank you for reviewing this solution! Feel free to fork this notebook, leave a comment, or connect for collaboration.*
